src/crypto/DeviceList.js

```diff
@@ -53,45 +53,30 @@ export default class DeviceList {
      * module:crypto/deviceinfo|DeviceInfo}.
      */
     downloadKeys(userIds, forceDownload) {
-        const self = this;
-
         // promises we need to wait for while the download happens
         const promises = [];
 
-        // list of userids we need to download keys for
-        let downloadUsers = [];
-
-        function perUserCatch(u) {
-            return function(e) {
-                console.warn('Error downloading keys for user ' + u + ':', e);
-            };
-        }
-
-        if (forceDownload) {
-            downloadUsers = userIds;
-        } else {
-            for (let i = 0; i < userIds.length; ++i) {
-                const u = userIds[i];
-
-                const inprogress = this._keyDownloadsInProgressByUser[u];
-                if (inprogress) {
-                    // wait for the download to complete
-                    promises.push(q.any(inprogress).catch(perUserCatch(u)));
-                } else if (!this.getStoredDevicesForUser(u)) {
-                    downloadUsers.push(u);
+        let needsFlush = false;
+        userIds.forEach((u) => {
+            if (this._keyDownloadsInProgressByUser[u]) {
+                // just wait for the existing download to complete
+                promises.push(this._keyDownloadsInProgressByUser[u]);
+            } else {
+                if (forceDownload || !this.getStoredDevicesForUser(u)) {
+                    this.invalidateUserDeviceList(u);
+                }
+                if (this._pendingUsersWithNewDevices[u]) {
+                    needsFlush = true;
                 }
             }
-        }
+        });
 
-        if (downloadUsers.length > 0) {
-            const r = this._doKeyDownloadForUsers(downloadUsers);
-            downloadUsers.map(function(u) {
-                promises.push(r[u].catch(perUserCatch(u)));
-            });
+        if (needsFlush) {
```

Comment:
every time i've read the flush code i've been confused, as flush to me always sounds as if you're trying to write to something and you want the writes to all complete.  But here, it's actually flushing *reads* to download new device lists, aiui?  I wonder if there is better naming to be had here...